# 4.8 Data Visualization with Python

# Contents:
1. Import libraries and data
2. Wrangling and Cleaning Customer Data
3. merging customer data with orders and products
4. Exporting Data

# 1. Import Libraries and Data

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# path to data folder
path = r'C:\Users\justi\OneDrive\Desktop\Class work\Achievement 4 Python (Instacart Basket Analysis)'

In [3]:
# import customer data set
customer = pd.read_csv(os.path.join(path, '02 data', 'original data', 'customers.csv'), index_col = False)

In [4]:
customer.shape

(206209, 10)

In [5]:
customer

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

# 2. Cleaning and wrangling customer data

In [6]:
# rename Surnam column 'Last Name' 
customer.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [7]:
customer

user_id First Name Last Name  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [8]:
# Remove fam_status column 
customer = customer.drop(columns = ['fam_status'])

In [9]:
customer

user_id First Name Last Name  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants  income  
0                  3  165665  
1                  0   59285  
2                  2   99568  
3                  0   42049  
4                  1   40374  
...              ...     ...  
206204             1  148828  
206205             3  168639  
206206             2   53374  
206207             1   99799  
206208             1   57095  

[206209 rows x 9 columns]

In [10]:
#Check
customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [11]:
#Check
customer.isnull().sum()

user_id             0
First Name      11259
Last Name           0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
income              0
dtype: int64

# There are 11,259 missing records in the 'First Name' column

In [12]:
# Checking NaN Records
customer_nan = customer[customer['First Name'].isnull() == True]

In [13]:
customer_nan

user_id First Name Last Name  Gender         STATE  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants  income  
53                 2   41709  
73                 0   82518  
82                 3  117099  
99                 1  155673  
105                3  151819  
...              ...     ...  
206038             3   87783  
206044             2  108488  
206090             1   45275  
206162             0   56325  
206171             2   59222  

[11259 rows x 9 columns]

# I will be removing the rows with NaN for first name since it is only 5% of the Data.

In [14]:
# Removing NaN records

customer_clean = customer[customer['First Name'].isnull() == False]

In [15]:
# Check
customer_clean

user_id First Name Last Name  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants  income  
0                  3  165665  
1                  0   59285  
2                  2   99568  
3                  0   42049  
4                  1   40374  
...              ...     ...  
206204             1  148828  
206205             3  168639  
206206             2   53374  
206207             1   99799  
206208             1   57095  

[194950 rows x 9 columns]

In [16]:
# Checking for duplicates
customer_dups = customer_clean[customer_clean.duplicated()]

In [17]:
customer_dups.shape

(0, 9)

In [18]:
customer_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, STATE, Age, date_joined, n_dependants, income]
Index: []

# There are no duplicates

In [19]:
for col in customer.columns.tolist():
 weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
if len(customer[weird]) > 0:
    print (col)

# There isn't any mixed-type data.

In [20]:
# Check
customer_clean.shape

(194950, 9)

In [21]:
# Import other data to combine with customer data
ords_prods = pd.read_pickle(os.path.join(path, '02 data', 'prepared data', 'orders_products_merged_with_flags.pkl'))

In [22]:
ords_prods.head()

Unnamed: 0_x  order_id  user_id  order_number orders_day_of_week  \
0             0   2539329        1             1             Monday   
1             1   2398795        1             2            Tuesday   
2             2    473747        1             3            Tuesday   
3             3   2254736        1             4          Wednesday   
4             4    431534        1             5          Wednesday   

   hour_of_day_ordered  days_since_prior_order  product_id  add_to_cart_order  \
0                    8                     NaN         196                  1   
1                    7                    15.0         196                  1   
2                   12                    21.0         196                  1   
3                    7                    29.0         196                  1   
4                   15                    28.0         196                  1   

   reordered  ...    price_range_loc         busiest day    busiest days  \
0          0  ...  Mid range product  Regularly Busy Day  Regularly Busy   
1          1  ...  Mid range product  Regularly Busy Day    Slowest Days   
2          1  ...  Mid range product  Regularly Busy Day    Slowest Days   
3          1  ...  Mid range product      Least Busy Day    Slowest Days   
4          1  ...  Mid range product      Least Busy Day    Slowest Days   

    Busiest hours  max_order  loyalty_flag avg_prices spending_flag  \
0  Average orders         10  New Customer   6.367797   Low Spender   
1  Average orders         10  New Customer   6.367797   Low Spender   
2     Most Orders         10  New Customer   6.367797   Low Spender   
3  Average orders         10  New Customer   6.367797   Low Spender   
4     Most Orders         10  New Customer   6.367797   Low Spender   

  order_frequency         frequency_flag  
0            20.5  Non-Frequent Customer  
1            20.5  Non-Frequent Customer  
2            20.5  Non-Frequent Customer  
3            20.5  Non-Frequent Customer  
4            20.5  Non-Frequent Customer  

[5 rows x 26 columns]

In [23]:
customer_clean.head()

user_id First Name Last Name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants  income  
0             3  165665  
1             0   59285  
2             2   99568  
3             0   42049  
4             1   40374

In [24]:
ords_prods.dtypes

Unnamed: 0_x                 int64
order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week          object
hour_of_day_ordered          int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
_merge                    category
Unnamed: 0_y                 int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
price_range_loc             object
busiest day                 object
busiest days                object
Busiest hours               object
max_order                    int64
loyalty_flag                object
avg_prices                 float64
spending_flag               object
order_frequency            float64
frequency_flag              object
dtype: object

In [25]:
customer_clean.dtypes

user_id          int64
First Name      object
Last Name       object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
income           int64
dtype: object

# 3. Merging Customer data with orders and products

In [26]:
# merging customer data with orders & products data using user_id as key

merged_data = customer_clean.merge(ords_prods, on = 'user_id', how = 'left')


In [27]:
merged_data.head(10)

user_id First Name Last Name  Gender     STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
5    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
6    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
7    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
8    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
9    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   n_dependants  income  Unnamed: 0_x  ...    price_range_loc  \
0             3  165665        443895  ...  Mid range product   
1             3  165665        443897  ...  Mid range product   
2             3  165665        443898  ...  Mid range product   
3             3  165665        443892  ...  Mid range product   
4             3  165665        443893  ...  Mid range product   
5             3  165665        443894  ...  Mid range product   
6             3  165665        443896  ...  Mid range product   
7             3  165665        443897  ...  Mid range product   
8             3  165665        443898  ...  Mid range product   
9             3  165665        443891  ...  Low range product   

          busiest day    busiest days   Busiest hours  max_order  \
0  Regularly Busy Day    Busiest Days  Average orders          8   
1  Regularly Busy Day  Regularly Busy     Most Orders          8   
2  Regularly Busy Day    Busiest Days     Most Orders          8   
3  Regularly Busy Day  Regularly Busy  Average orders          8   
4  Regularly Busy Day    Slowest Days     Most Orders          8   
5  Regularly Busy Day    Busiest Days  Average orders          8   
6  Regularly Busy Day    Busiest Days  Average orders          8   
7  Regularly Busy Day  Regularly Busy     Most Orders          8   
8  Regularly Busy Day    Busiest Days     Most Orders          8   
9  Regularly Busy Day  Regularly Busy  Average orders          8   

   loyalty_flag  avg_prices  spending_flag order_frequency    frequency_flag  
0  New Customer    7.988889    Low Spender            19.0  Regular Customer  
1  New Customer    7.988889    Low Spender            19.0  Regular Customer  
2  New Customer    7.988889    Low Spender            19.0  Regular Customer  
3  New Customer    7.988889    Low Spender            19.0  Regular Customer  
4  New Customer    7.988889    Low Spender            19.0  Regular Customer  
5  New Customer    7.988889    Low Spender            19.0  Regular Customer  
6  New Customer    7.988889    Low Spender            19.0  Regular Customer  
7  New Customer    7.988889    Low Spender            19.0  Regular Customer  
8  New Customer    7.988889    Low Spender            19.0  Regular Customer  
9  New Customer    7.988889    Low Spender            19.0  Regular Customer  

[10 rows x 34 columns]

In [28]:
# Check
merged_data.shape

(30629741, 34)

# 4. Exporting Merged Data set

In [29]:
# Export newly merged data
merged_data.to_pickle(os.path.join(path, '02 data', 'prepared data', 'Data Merged.pkl'))

In [30]:
# Export Cleaned customer data
customer_clean.to_csv(os.path.join(path, '02 data', 'prepared data', 'customer_clean.csv'))